core/src/main/java/io/undertow/server/handlers/ratelimit/BitShiftSingleWindowRateLimiter.java

```diff
@@ -0,0 +1,208 @@
+
+        final String ipAddress = getIPAddress(exchange, true);
+        final String key = generateKey(ipAddress);
+        AtomicInteger ai = requestCounter.computeIfAbsent(key, v -> new AtomicInteger());
+        return ai.accumulateAndGet(1, (value, upTick) -> {
+            // JIC
+            if (value < TICK_BORDER) {
+                return value + upTick; //uptick = +1
+            } else {
+                return Integer.MAX_VALUE;
+            }
+        });
+    }
+
+    @Override
+    public int current(HttpServerExchange exchange) {
+        final String ipAddress = getIPAddress(exchange, true);
+        final String key = generateKey(ipAddress);
+        AtomicInteger entry = requestCounter.get(key);
+        if(entry != null) {
+            return entry.get();
+        } else {
+            return -1;
+        }
+    }
+
+    private String getIPAddress(final HttpServerExchange exchange, final boolean warn) {
+        final InetSocketAddress sourceAddress = exchange.getSourceAddress();
+        InetAddress address = sourceAddress.getAddress();
+        if (address == null) {
+            // this can happen when we have an unresolved X-forwarded-for address
+            // in this case we just return the IP of the balancer
+            //TODO: this needs impr
+            address = ((InetSocketAddress) exchange.getConnection().getPeerAddress()).getAddress();
+            if(warn) {
+                UndertowLogger.REQUEST_LOGGER.rateLimitFailedToGetProperAddress(exchange.getRequestURI(), address.getHostAddress());
+            }
+        }
+        return address.getHostAddress();
+    }
+
+    @Override
+    public String getLimiterID() {
+        return "bit-shift-window";
+    }
+
+    @Override
+    public RateLimitUnit getUnit() {
+        return RateLimitUnit.REQUEST;
+    }
+
+    private void evictionCheck(HttpServerExchange exchange) {
+        // we need to parasite on IO threads for eviction.
+        XnioExecutor.Key key = this.evictionKey;
+        if (key == null) {
+            this.evictionKey = WorkerUtils.executeAfter(exchange.getIoThread(), new Runnable() {
+                @Override
+                public void run() {
+                    evictionKey = null;
+                    evictOldWindow();
+                }
+            }, this.windowDuration, TimeUnit.SECONDS);
+        }
+    }
+
+    private void evictOldWindow() {
+        // evict entries that are not 'current' or 'current+1' - just in case eviction starts in one window and finish in next.
+        // in such case 'current' will become stale already and will be evicted on next call. Thats fine.
+        // prefix + 1 will translate into bitshift ++
+        final long currentPrefix = generatePrefix();
+        final String current = String.valueOf(currentPrefix);
+        final String next = String.valueOf(currentPrefix + 1);
+
+        ConcurrentHashMap.KeySetView<String, AtomicInteger> keys = requestCounter.keySet();
```

Comment:
String could be turned into Long with high bits holding IP in pure form and lower working as in this case. But its harder to follow.